core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -650,14 +650,24 @@ private IndexResult innerIndex(Index index) throws IOException {
                 }
                 updatedVersion = index.versionType().updateVersion(currentVersion, expectedVersion);
                 index.parsedDoc().version().setLongValue(updatedVersion);
+
+                // Update the document's sequence number and primary term, the
+                // sequence number here is derived here from either the sequence
+                // number service if this is on the primary, or the existing
+                // document's sequence number if this is on the replica. The
+                // primary term here has already been set, see
+                // IndexShard.prepareIndex where the Engine.Index operation is
+                // created
+                index.parsedDoc().updateSeqID(seqNo, index.primaryTerm());
+
                 if (currentVersion == Versions.NOT_FOUND && forceUpdateDocument == false) {
                     // document does not exists, we can optimize for create, but double check if assertions are running
                     assert assertDocDoesNotExist(index, canOptimizeAddDocument == false);
                     index(index.docs(), indexWriter);
                 } else {
                     update(index.uid(), index.docs(), indexWriter);
                 }
-                indexResult = new IndexResult(updatedVersion, seqNo, deleted);
+                indexResult = new IndexResult(updatedVersion, seqNo, index.primaryTerm(), deleted);
```

Comment:
I included it because the sequence number is included in the result. Also, it's used when constructing a new `Index` op from an `Engine.IndexResult`:

```java
        public Index(Engine.Index index, Engine.IndexResult indexResult) {
            this.id = index.id();
            this.type = index.type();
            this.source = index.source();
            this.routing = index.routing();
            this.parent = index.parent();
            this.seqNo = indexResult.getSeqNo();             // <-- here
            this.primaryTerm = indexResult.getPrimaryTerm(); // <-- and here
            this.version = indexResult.getVersion();
            this.versionType = index.versionType();
            this.autoGeneratedIdTimestamp = index.getAutoGeneratedIdTimestamp();
        }
```

(Also used when creating a new `Delete` op from a `DeleteResult`